src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs

```diff
@@ -230,9 +230,28 @@ private void ResolveDotNetCoreAsymmetricAlgorithm(AsymmetricSecurityKey key, str
             if (ecdsaKey != null)
             {
                 if (ecdsaKey.ECDsa != null)
+                {
                     _ecdsaCng = ecdsaKey.ECDsa;
+                    return;
+                }
+            }
+
+            JsonWebKey webKey = key as JsonWebKey;
+            if (webKey.Kty == JsonWebAlgorithmsKeyTypes.RSA)
+            {
+                RSAParameters parameters = CreateRsaParametersFromJsonWebKey(webKey, willCreateSignatures);
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    _rsa = new RSACng();
                 else
-                    _ecdsaCng = new ECDsaCng(ecdsaKey.CngKey);
+                    _rsa = new RSAOpenSsl();
+
+                _rsa.ImportParameters(parameters);
+                _disposeRsa = true;
+                return;
+            }
+            else if (webKey.Kty == JsonWebAlgorithmsKeyTypes.EllipticCurve)
+            {
+                CreateECDsaFromJsonWebKey(webKey, willCreateSignatures);
```

Comment:
This code path is CNG-specific, it will crash on other platforms. You absolutely need to ensure we're on Windows at this point.